### ISSUE TITLE
Weird points about the code

### ISSUE THREAD

**Issue by lyxok1** [TKKim93/APE#5]
Thanks for sharing the code. I am currently trying to reimplement the code and find some wierd points about the code. Here is the point

1.  Line 231 in main.py
`
perturb, clean_vat_logits = P(target_data)
`
The variable `clean_val_logits` logically should be the predict logits from unperturbed target data, while from the forward defenition of `P` (Line 55-77 in perturb.py)

```python
  def forward(self, inputs):
          with disable_tracking_bn_stats(self.feature_extractor):
              with disable_tracking_bn_stats(self.classifier):
                  features = self.feature_extractor(inputs)
                  logits = self.classifier(features)[1].detach()
                  # prepare random unit tensor
                  d = l2_normalize(torch.randn_like(inputs).to(inputs.device))
                  # calc adversarial direction
                  x_hat = inputs
                  x_hat = x_hat + self.xi * d
                  x_hat.requires_grad = True
                  features_hat = self.feature_extractor(x_hat)
                  logits_hat = self.classifier(features_hat, reverse=True, eta=1)[1]
                  prob_hat = F.softmax(logits_hat, 1)
                  adv_distance = (prob_hat * torch.log(1e-4 + prob_hat)).sum(1).mean()
                  adv_distance.backward()
                  d = l2_normalize(x_hat.grad)
                  self.feature_extractor.zero_grad()
                  self.classifier.zero_grad()
                  r_adv = d * self.eps
                  return r_adv.detach(), features
```
We see the second return value is the `features` extract from backbone network, which is not the logits val, and can even be inconsistent with the logits in dimension. Similar problem happened in (Line 234 main.py)
```python
perturb_logits = F1(perturb_features)[0]
```
From Line 75-89 in basenet.py
```python
class Predictor_deep_latent(nn.Module):
    def __init__(self, num_class=64, inc=4096, temp=0.05):
        super(Predictor_deep_latent, self).__init__()
        self.fc1 = nn.Linear(inc, 512)
        self.fc2 = nn.Linear(512, num_class, bias=False)
        self.num_class = num_class
        self.temp = temp

    def forward(self, x, reverse=False, eta=0.1):
        x = self.fc1(x)
        if reverse:
            x = grad_reverse(x, eta)
        x = F.normalize(x)
        x_out = self.fc2(x) / self.temp
        return x, x_out
```
We see the `F1(perturb_features)[0]` will return `x` not `x_out`, i.e. not the classification output, this is quite different from the Eq (4) in the paper.

2. Another wierd point also comes from the codeblock in  Line 75-89 in basenet.py as shown above, the returned `x` value is the one passed through `grad_reverse` function, however, in Line 205-212 main.py, it is used to calculate MMD loss
```python
        latent_F1, out1 = F1(output)
        latent_F1_tu, out_F1_tu = F1(output_tu)

        # supervision loss
        loss = criterion(out1, target)

        # attraction scheme
        loss_msda = 10 * mmd.mix_rbf_mmd2(latent_F1, latent_F1_tu, sigma)
```
Because of the `grad_reverse` function, when minimizing `loss_msda`, the backbone part is optimized toward the direction that maximize the MMD distance between two subdistribution ? This is wired and seems not consisitent with the insight behind the attraction scheme in paper.

3. The third wierd point lies in the perturbation function (also from Line 55-77 of perturb.py), the paper says first prototype (the final classification layer) is perturbed then noise is generate, but it seems the noise is generated directly based on current prototype value and entropy minimization.

Great appreciation if you can provide some information about the questions above
